### PR TITLE
Review

### PR DESCRIPTION
**Pull request by srush**
<a href='#crh-start'></a><a href='#crh-data-%7B%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
<a href='https://www.codereviewhub.com/srush/nn_coref/pull/1?approve=1'></a>&nbsp;<a href='https://github.com/srush/nn_coref/pull/1'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by srush** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/srush/nn_coref/pull/1'>Refresh</a>.<a href='#crh-update-none'></a>

File: do_everything.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        sys.stdout.flush()
+    else:
+        out = sub.check_output("reference-coreference-scorers/v8.01/scorer.pl"\
+        " all test.key nn/conllouts/load_and_pred.bps.out none", shell=True)
+        print "conll scorer output:\n\n%s" % out  
+        sys.stdout.flush()      
+    
+def main():
+    """
+    make sure to run this script from the main nn_coref/ directory
```
Comment (srush) [inline do_everything.py:188]: General Style: Comments as full sentences, capital and period.

File: do_everything.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+AN_PATH = "animate.unigrams.txt"
+INAN_PATH = "inanimate.unigrams.txt"
+FEATS = "FINAL+MOARANAPH+MOARPW+bilexical"
+
+# hdf5 conversion
+NUM_CONVERT_PROCS = 30
+FEAT_SET_ABBREV = "basicp"
+
+# training
+PRETRAIN = True
```
Comment (srush) [inline do_everything.py:31]: Can you make some of these options with argparse? Seems like you may want to call with pretrain = False

File: do_everything.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
+#!/bin/python
```
Comment (srush) [inline do_everything.py:1]: do_everything.py is an intense name. Maybe call this train.py and make it a bit more extensible (see other comments).

File: do_everything.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+import os
+import shutil
+import subprocess as sub
+import sys
+
+# This script is intended to show how to reproduce the experiments in the paper,
```
Comment (srush) [inline do_everything.py:8]: Make this a docstring """ instead of #

File: do_everything.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+def text_feats_to_hdf5():
+    """
+    assumes in nn/ directory
+    """
+    tr_cmd = "python text_feats_to_hdf5.py -n %d ../NONE-%s-%s %s %s"
+    dev_cmd = "python text_feats_to_hdf5.py -n %d -t ../NONE-%s-%s %s %s"
+   
+    # convert pairwise train features
+    sub.call(tr_cmd % (NUM_CONVERT_PROCS,FEATS,"pwTrainFeats.txt","train_"+FEAT_SET_ABBREV,"pw"), shell=True)
```
Comment (srush) [inline do_everything.py:86]: Minor: wrap this a call function instead of passing shel=True every time.

File: do_everything.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+                      "train_"+FEAT_SET_ABBREV, "train_"+FEAT_SET_ABBREV, "-random_init")
+        print "running: %s" % cmd
+        sys.stdout.flush()
+        sub.call(cmd,shell=True) 
+
+def predict(pfx="dev_"):
+    tmplt = "th %s -pwDevFeatPfx %s -anaDevFeatPfx %s -load_and_predict %s %s"
+    load_pw = ("-pwFullSerFi models/%s-0.500000-1.200000-1.000000.model-full_%s_%s-pw" 
+                % (FEAT_SET_ABBREV,"g1" if FULL_MODEL == "g1" else "g2","PT" if PRETRAIN else "RI"))
+    load_ana = ("-anaFullSerFi models/%s-0.500000-1.200000-1.000000.model-full_%s_%s-na"
```
Comment (srush) [inline do_everything.py:154]: Whenever possible pull these magic numbers constants out into constants. Not sure where they are coming from (default args)?

File: do_everything.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+                  pfx+FEAT_SET_ABBREV,pfx+FEAT_SET_ABBREV,load_pw,load_ana)
+    print "running: " % cmd
+    sys.stdout.flush()
+    sub.call(cmd,shell=True)
+
+def get_conll_fmt_output(dev=True):
+    """
+    assumes in modifiedBCS/ directory
+    """
+    if dev:
```
Comment (srush) [inline do_everything.py:166]: Try not to repeat strings like this. Just "if" the arg. (and below).

File: nn/ana_model.lua [hunk 8]
```diff
@@ -0,0 +1,251 @@
+require 'nn'
+dofile('init.lua')
+require 'cmd_adagrad'
+require 'coref_utils'
+require 'sparse_doc_data'
+
+ANAModel = {}
+ANAModel.__index = ANAModel
```
Comment (srush) [inline nn/ana_model.lua:8]: Use torch.Class for the object stuff, gets rid of the metatable nonsense. (sorry I was wrong here before)

File: nn/ana_model.lua [hunk 9]
```diff
@@ -0,0 +1,251 @@
+require 'nn'
+dofile('init.lua')
```
Comment (srush) [inline nn/ana_model.lua:2]: This is a bit nonstandard. Is it needed?

File: nn/ana_model.lua [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+
+function ANAModel.make(naD, hiddenUnary, c0, c1) 
+    model = {}
+    setmetatable(model,ANAModel)
+    model.hiddenUnary = hiddenUnary
+    model.c0 = c0
+    model.c1 = c1
+
+    local naNet = nn.Sequential()
+    naNet:add(nn.LookupTable(naD,hiddenUnary))
```
Comment (srush) [inline nn/ana_model.lua:18]: Minor Style: always period after ,

File: nn/ana_model.lua [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+    model.hiddenUnary = naNet:get(1).weight:size(2)
+    model.naNet = naNet
+    
+    model.nastates = {{},{},{},{}}
+    
+    collectgarbage()
+    return model
+end 
+
+-- this just gets the delta for each datapt based on a slack rescaled (binary) hinge loss
```
Comment (srush) [inline nn/ana_model.lua:52]: Can you improve this comment a bit? Roughly what are the arguments and what does it return?

File: nn/ana_model.lua [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+   end
+   return delt
+end
+
+
+--[[ 
+   - computes gradients assuming parameters have been zeroed out
+   - naData is an SpDMData
+--]]    
+function ANAModel:docGrad(d,naData,targets)
```
Comment (srush) [inline nn/ana_model.lua:76]: Can this just be miniBatchGrad? Doesn't seem crucial that this code knows what a doc is.

File: nn/ana_model.lua [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+       -- use document sized minibatches
+       for d = 1, naTrData.numDocs do
+          if d % 200 == 0 then
+             print("doc " .. tostring(d))
+             collectgarbage()
+          end 
+
+          self.naNet:zeroGradParameters()
+          self:docGrad(d,naTrData,trTargets)
+          cmdAdaGradC(self.naNet:get(1).weight, self.naNet:get(1).gradWeight,
```
Comment (srush) [inline nn/ana_model.lua:109]: This code is a bit scary and bug prone. I'd recommend naming the units (self.lookup vs self.naNet:get(1)) before you add them to the network. self.lookup = nn.Lookup() nn:add(self.lookup). That way if the network changes things dont get out of order.

File: nn/ana_model.lua [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+                      eta, lamb, self.nastates[1])
+          cmdAdaGradC(self.naNet:get(3).bias, self.naNet:get(3).gradBias, 
+                      eta, lamb, self.nastates[2])  
+          cmdAdaGradC(self.naNet:get(5).weight, self.naNet:get(5).gradWeight, 
+                      eta, lamb, self.nastates[3])                         
+          cmdAdaGradC(self.naNet:get(5).bias, self.naNet:get(5).gradBias, 
+                      eta, lamb, self.nastates[4])
```
Comment (srush) [inline nn/ana_model.lua:116]: If possible, this should be a loop. maybe in addition to self.nastates[i] have self.weights[i] and self.grads[i].

File: nn/ana_model.lua [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+    return targets   
+end
+
+cmd = torch.CmdLine()
+cmd:text()
+cmd:text()
+cmd:text('Training anaph model')
+cmd:text()
+cmd:text('Options')
+cmd:option('-hidden_unary', 128, 'Hidden layer size')
```
Comment (srush) [inline nn/ana_model.lua:208]: Clean up the writing here. Roughly this is no less important than the paper (and maybe more in practice). 
- Consistent capitalization 
- Consistent punctuation. 
- Very clear option names: "Fi"?  "Pfx"?
- I would recommend clear non-math names '-adagrad-rate', 0.1, "AdaGrad learning rate (\eta).", although that is a style choice.

File: nn/ante_model.lua [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+        if clusts[d]:anaphoric(m) then        
+            local start = ((m-1)*m)/2 -- one behind first pair for mention m
+         
+         -- score all the potential antecedents
+            for a = 1, m-1 do
+                scores[start+a] = self.pwNet:forward(pwData:getFeats(d,m,a))
+            end
+            
+            local late = maxGoldAnt(clusts[d],scores,m,start)
+            scores[start+m] = -math.huge -- this is a hack; the method below looks thru m
```
Comment (srush) [inline nn/ante_model.lua:76]: Comments on their own line.

File: nn/ante_model.lua [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+       for d = 1, pwData.numDocs do
+          if d % 200 == 0 then
+             print("doc " .. tostring(d))
+             collectgarbage()
+          end 
+          local pwnz = {}
+          self.pwNet:zeroGradParameters()
+          self:docGrad(d,pwData,clusts,pwnz)
+          -- do pw gradients
+          colTblCmdAdaGrad(self.pwNet:get(1).weight, self.pwNet:get(1).gradWeight, pwnz,
```
Comment (srush) [inline nn/ante_model.lua:139]: Similar comments as in Ante. Ideally this training function could be shared between the two models.

File: nn/cmd_adagrad.lua [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+- `config.eta` : learning rate
+- `state` : a table describing the state of the optimizer; after each
+         call the state is modified
+- `state.paramVariance` : vector of temporal variances of parameters
+
+RETURN:
+- `x` : the new x vector
+- `f(x)` : the function, evaluated before the update
+]]
+function cmdAdaGrad(opfunc, x, config, state)
```
Comment (srush) [inline nn/cmd_adagrad.lua:25]: Style: Can you make these functions under a namespace CMD. I.e. CMD = {} function CMD.AdaGrad() etc. Benefits: 1) functions are non-global 2) clear where they come from. 

Alternatively, you can put it in optim's namespace if that makes more sense.

File: nn/cmd_adagrad.lua [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+   x:cmul(dfdx) -- recall that dfdx stores the sign now
+   
+   -- (6) update evaluation counter
+   state.evalCounter = state.evalCounter + 1
+
+   -- return x*, f(x) before optimization
+   return x,{fx}
+end
+
+--[[ all the following functions assume the input (x) and the subgradient (g) are contiguous
```
Comment (srush) [inline nn/cmd_adagrad.lua:73]: Can you assert this in the code? There should be a is_contiguous torch function.

File: nn/coref_utils.lua [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+            local perm = torch.randperm(m)
+            local r = torch.randn(numNZ)*scale
+            for i = 1, numNZ do
+                W[perm[i]][j] = r[i]
+            end
+        end
+    end
+end
+
+function checkContigAndSutsInit(net,numNZ)
```
Comment (srush) [inline nn/coref_utils.lua:108]: Does this have anything to do with coref? Maybe a separate file for model init functions.

File: nn/coref_utils.lua [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+        if mod.bias then
+            assert(mod.bias:isContiguous())
+        end
+    end
+end
+
+
+--[[ misc --]]
+
+function setFromList(tbl)
```
Comment (srush) [inline nn/coref_utils.lua:140]: This also seems different. Maybe a general utils file.

File: nn/coref_utils.lua [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+    local d = 1
+    for line in io.lines(fi) do
+        table.insert(OPCs, OPClust.makeFromStr2(line,spDocData:numMents(d)))
+        d = d+1
+    end
+    collectgarbage()
+    return OPCs
+end
+
+function maxGoldAnt(clust,scoreBuf,ment,start)
```
Comment (srush) [inline nn/coref_utils.lua:66]: If this is non-global use "local function".

File: nn/cmd_adagrad.lua [hunk 23]
```diff
@@ -0,0 +1,173 @@
+--[[
+   Composite Mirror Descent (with L1 regularization) AdaGrad update.
+   Several implementations of equation (25) in http://www.cs.berkeley.edu/~jduchi/projects/DuchiHaSi10.pdf
```
Comment (srush) [inline nn/cmd_adagrad.lua:3]: Give the paper cite here as well (google-style does this, kind of like it. )

**Comment by srush** [comment]
Some comments to start with.
